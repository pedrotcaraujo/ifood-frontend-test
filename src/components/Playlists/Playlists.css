.Playlists {
    width: 100%;
    margin-top: 3em;
}

.Playlists-items {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    margin: 0;
    padding: 0;
}

.Playlists-item {
    list-style: none;
    margin: 1em;
}

.Playlists-figure {
    margin: 0;
}

.Playlists-message {
    font-size: 2em;
    text-shadow: 1px 1px 5px #000;
}

.Playlists-player {
    position: relative;
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
}

.Playlists-player:hover::after {
    content: ' ';
    position: absolute;
    z-index: 1;
    background: #000;
    height: 101%;
    width: 101%;
    opacity: .8;
    top: -.5px;
    left: -0.5px;
}

.Playlists-player:hover .Playlists-button {
    display: block;
}

.Playlists-image {
    box-shadow: 1px 1px 50px #000;
}

.Playlists-button {
    display: none;
    position: absolute;
    z-index: 2;
    border: 0;
    background: transparent;
    box-sizing: border-box;
    width: 0;
    height: 74px;
  
    border-color: transparent transparent transparent #fff;
    transition: 100ms all ease;
    cursor: pointer;
  
    border-style: solid;
    border-width: 37px 0 37px 60px;
}

.Playlists-figcaption {
    margin-top: 10px;
}