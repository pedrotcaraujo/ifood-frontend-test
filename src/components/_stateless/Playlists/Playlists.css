.Playlists {
    width: 100%;
    margin-top: 3em;
}

.Playlists-items {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    margin: 0;
    padding: 0;
}

.Playlists-item {
    list-style: none;
    margin: 1em;
}

.Playlists-message {
    font-size: 2em;
    text-shadow: 1px 1px 5px #000;
}