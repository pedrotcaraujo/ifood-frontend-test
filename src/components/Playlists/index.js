import React, { Component } from 'react';
import axios from 'axios';
import { removeEmpty, fuzzySearch } from '../../utils';
import './Playlists.css';

import FeaturedPlaylistDispatcher from '../../dispatchers/FeaturedPlaylistDispatcher';
import PlaylistsConstants from '../../constants/PlaylistsConstants';
import FilterStore from '../../stores/FilterStore';

import Loader from '../Loader';

const URL = 'https://api.spotify.com/v1/browse/featured-playlists';
const PULLING_TIMEOUT = 300000;

class Playlists extends Component {    
    state = { loaded: false, search: '', message: null, playlists: { items: [] }, playing: {} }
    timeouts = []

    componentDidMount() {
        FilterStore.addListener(this.onFilterChange)
        if (this.props.token) {
            window.requestAnimationFrame(() => this.update(this));
        }
    }

    componentWillUnmount() {
        FilterStore.removeListener(this.onFilterChange)
        window.clearTimeout(this.timeout);
    }

    update(context, params = {}) {
        context.setState({ loaded: false })
        axios.get(URL, {
            params,
            headers: {
              'Authorization': 'Bearer ' + context.props.token
            }
        })
        .then(({ data }) => {
            context.setState({ loaded: true, ...data });
            context.pulling(PULLING_TIMEOUT, () => context.update(context, params));
        })
        .catch(() => {
            context.setState({ 
                loaded: true, 
                message: 'Não foi possível fazer a busca, tente novamente...',
                playlists: { 
                    items: [] 
                }
            })
        });
    }

    pulling(timeout, callback) {
        this.timeouts.push(window.setTimeout(() => window.requestAnimationFrame(callback), timeout));
    }

    onFilterChange = () => {
        const { current, data } = FilterStore.getState();
        const { search, ...params } = data;

        if (current.name === 'search') {
            this.setState(Object.assign({...this.state}, { search: current.value }))
            return;
         }

        if (this.timeouts.length > 0) {
            this.timeouts.forEach(timeout => {
                window.clearTimeout(timeout)
            })
        }

        if (this.state.loaded) {
            this.update(this, removeEmpty(params));
        }
    }

    onPlay(playlist) {
        FeaturedPlaylistDispatcher.dispatch({
            type: PlaylistsConstants.UPDATE,
            data: Object.assign({}, { 
                playing: playlist,
                playlists: this.state.playlists
            })
        })
    }

    render() {
        return (
            <Loader loaded={this.state.loaded}>
                <div className="Playlists">
                    <h2 className="Playlists-message">{this.state.message}</h2>
                    <ul className="Playlists-items">
                    {this.state.playlists.items.filter(item => fuzzySearch(item.name, this.state.search)).map(item => (
                        <li onClick={() => this.onPlay(item)} className="Playlists-item" key={item.id}>
                            <figure className="Playlists-figure">
                                <div className="Playlists-player">
                                    <button className="Playlists-button"/>
                                    <img className="Playlists-image" src={item.images[0].url} alt={item.name}/>
                                </div>
                                <figcaption className="Playlists-figcaption">{item.name}</figcaption>
                            </figure>
                        </li>                    
                    ))}
                    </ul>
  
                </div>
            </Loader>
        )
    }
}

export default Playlists;