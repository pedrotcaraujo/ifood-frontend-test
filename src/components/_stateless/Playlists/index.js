import React from 'react';
import PropTypes from 'prop-types';
import Playlist from '../Playlist';
import fuzzySearch from '../../../utils/fuzzySearch';
import './Playlists.css';

const Playlists = ({data, search, onPlay}) => (
    <div className="Playlists">
        <h2 className="Playlists-message">{data.message}</h2>
        <ul className="Playlists-items">
        {data.playlists.items.filter(item => fuzzySearch(item.name, search)).map(item => (
            <li onClick={() => onPlay(item)} className="Playlists-item" key={item.id}>
                <Playlist name={item.name} image={item.images[0].url}/>
            </li>                    
        ))}
        </ul>

    </div>
);

Playlists.propTypes = {
    data: PropTypes.object.isRequired,
    search: PropTypes.string,
    onPlay: PropTypes.func
};


export default Playlists;