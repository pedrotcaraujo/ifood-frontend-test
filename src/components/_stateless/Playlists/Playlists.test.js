import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Playlists from './index';

it('renders correctly', () => {
  const tree = renderer
    .create(<Playlists data={{ message: 'Teste', playlists: { items: [] }}} />)
    .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Playlists data={{ message: 'Teste', playlists: { items: [] }}} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
