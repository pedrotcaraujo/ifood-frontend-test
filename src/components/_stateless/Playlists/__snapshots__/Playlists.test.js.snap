// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Playlists"
>
  <h2
    className="Playlists-message"
  >
    Teste
  </h2>
  <ul
    className="Playlists-items"
  />
</div>
`;
